File: literature-app/src/api/workspace.ts
import { tauriInvoke } from './client';
import type { Workspace } from '../types';

export const listWorkspaces = () => tauriInvoke<Workspace[]>('list_workspaces');
export const createWorkspace = (data: { name: string; description?: string }) =>
  tauriInvoke<Workspace>('create_workspace', { req: data });
export const getWorkspace = (id: string) => tauriInvoke<Workspace>('get_workspace', { id });
export const updateWorkspace = (id: string, data: { name?: string; description?: string }) =>
  tauriInvoke<Workspace>('update_workspace', { id, req: data });
export const deleteWorkspace = (id: string) => tauriInvoke<boolean>('delete_workspace', { id });

File: backend/src/routes/workspace.rs
use axum::{extract::{Path, State}, Json};
use neo4rs::Graph;
use crate::errors::AppError;
use crate::models::dto::{CreateWorkspaceRequest, DeleteResponse, UpdateWorkspaceRequest};
use crate::models::workspace::Workspace;
use crate::repositories::neo4j_repo::Neo4jRepo;
use crate::services::workspace::WorkspaceService;

pub async fn create_workspace(
    State(graph): State<Graph>,
    Json(req): Json<CreateWorkspaceRequest>,
) -> Result<Json<Workspace>, AppError> {
    let repo = Neo4jRepo::new(graph);
    let workspace = WorkspaceService::create(&repo, req.name, req.description).await?;
    Ok(Json(workspace))
}

pub async fn list_workspaces(
    State(graph): State<Graph>,
) -> Result<Json<Vec<Workspace>>, AppError> {
    let repo = Neo4jRepo::new(graph);
    let workspaces = WorkspaceService::list(&repo).await?;
    Ok(Json(workspaces))
}

pub async fn get_workspace(
    State(graph): State<Graph>,
    Path(id): Path<String>,
) -> Result<Json<Workspace>, AppError> {
    let repo = Neo4jRepo::new(graph);
    let workspace = WorkspaceService::get(&repo, &id).await?;
    Ok(Json(workspace))
}

pub async fn update_workspace(
    State(graph): State<Graph>,
    Path(id): Path<String>,
    Json(req): Json<UpdateWorkspaceRequest>,
) -> Result<Json<Workspace>, AppError> {
    let repo = Neo4jRepo::new(graph);
    let workspace = WorkspaceService::update(&repo, &id, req.name, req.description).await?;
    Ok(Json(workspace))
}

pub async fn delete_workspace(
    State(graph): State<Graph>,
    Path(id): Path<String>,
) -> Result<Json<DeleteResponse>, AppError> {
    let repo = Neo4jRepo::new(graph);
    WorkspaceService::delete(&repo, &id).await?;
    Ok(Json(DeleteResponse { deleted: true }))
}
